### PR TITLE
Feature/swagger

### PR DESCRIPTION
**Pull request by rflechner**
Hello,

In this PR I propose a solution to generate a swagger for Giraffe. 
Issue https://github.com/giraffe-fsharp/Giraffe/issues/79 has label `help wanted` 😃  .

[My solution for Suave](https://rflechner.github.io/Suave.Swagger/) was effectively not really easy to use.
Documentation and service implementation were too strongly coupled and the DSL was really verbose.

The good news is that we still have to declare our API routes the same way as before but to enable the route analysis we have to surround the app declaration with quotation marks.

With that in place we can decouple the app declaration from the analysis required to generate the swagger documentation. In other words this solution has the avantage to avoid corrupting your service implementation.

## How does it work ?

I introduced the `documents` function that takes two arguments:
1. the quotation expression containing webservice implementation.
2. a `DocumentationConfig` argument.

This function does the analysis of your quotation to generate Swagger documentation.

`DocumentationConfig` contains the following properties:

- `MethodCallRules`: allow you to provide custom functions to enrich DSL and / or quotation analysis.
- `DocumentationAddendums`: allow you to add more informations to the documentation without introducing service implementation modification.

I introduced `==>` operator that gives the possibility to add `decorations` in routes implementations.

## Examples

There are 2 solutions to add documentation for a route.

[See example](https://github.com/rflechner/Giraffe/blob/feature/swagger/samples/SwaggerSample/Program.fs#L84)

- [First one](https://github.com/rflechner/Giraffe/blob/feature/swagger/samples/SwaggerSample/Program.fs#L102)

```fsharp
...
operationId "send_a_car" ==>
	consumes tcar ==>
		produces typeof<Car> ==>
			route "/car2" >=> submitCar
...
```

- [Second one](https://github.com/rflechner/Giraffe/blob/feature/swagger/samples/SwaggerSample/Program.fs#L138)

using `DocumentationAddendums`

```fsharp
...
route "/car" >=> submitCar
...

let docAddendums =
    fun (route:Analyzer.RouteInfos) (path:string,verb:HttpVerb,pathDef:PathDefinition) ->
        match path,verb,pathDef with
        | "/car", HttpVerb.Post,def ->
            let ndef = 
                (def.AddConsume "model" "application/json" Body typeof<Car>)
                    .AddResponse 200 "application/json" "A car" typeof<Car>
            path, verb, ndef
...
```

PS: I added a SLN file to improve experience with Jetbrains Rider IDE.
I propose to create a separated NuGet for `Giraffe.Swagger`.
In futur, SwaggerUi could be a submodule of the repository (if you like and accept this PR 😄  ).

Some features could be missing and some quotations could be difficult to parse.
For the moment, analyzer works with most basics [default httphandlers](https://github.com/giraffe-fsharp/Giraffe#default-httphandlers).

I only implemented:

- GET
- POST
- PUT
- PATCH
- DELETE
- route
- routeCi
- routef
- setStatusCode
- text
- json
- choose
- subRouteCi
- subRoute

You can build and run [SwaggerSample/Program.fs](https://github.com/rflechner/Giraffe/blob/feature/swagger/samples/SwaggerSample/Program.fs) and 
 go to http://localhost:5000/swaggerui/

screen_giraffe_swagger1

Regards.

@rflechner

### CODE REVIEW

**Comment by dsyme** [comment]
I only glanced at this - this is really cool.  One thing: could you use ``ReflectedDefinition`` and place the quotation here?
```
    documents <@ documentMap @> ...
```
I just wondered that when I looked - there might be good reasons for not doing it this way.

**Comment by rflechner** [comment]
Hi,


Thank you very much for your review.


I tried to use `[<ReflectedDefinition(true)>]` in `documents` signature like this:

```fsharp

// in lib

let documents ([<ReflectedDefinition(true)>] webapp:Expr<HttpFunc -> HttpContext -> HttpFuncResult>) (configuration:DocumentationConfig->DocumentationConfig) =
	...

// in app
let documentedApp = choose [....]

documents <@ documentApp @> ...
```

`ReflectedDefinitionAttribute` is effectively a really good solution to avoid `QuotationEvaluator.Evaluate webapp`.
But I have a problem to discover routes because expression is:

```fsharp
WithValue (<fun:documentedApp@69-34>, PropertyGet (None, documentedApp, []))
```

Did I misunderstand something ?

Regards,

@rflechner

**Comment by dustinmoris** [comment]
Hey, thank you so much for this work! Looks really interesting and I'll review it in more detail over the next few days now that I finished the work for the `1.0.0` release of Giraffe!

**Comment by cotyar** [comment]
After updating to Giraffe 1.0.0 the sample is failing to start with Quotation.Evaluator throwing saying it's called with parameter of incorrect type.

**Comment by rflechner** [comment]
Hi, 


I merged develop into my branch.


`QuotationEvaluator` was failing for `route  "/everytime"  >=> warbler (fun _ -> (time() |> text))`.

I found a solution with `[<ReflectedDefinition(true)>]` as suggested by @dsyme.

`ReflectedDefinitionAttribute` seem to work only with classes.
So I created `swaggerOf` type.


The new syntax to declare a swagger is: 

```F#
let webApp =
    swaggerOf
        (choose [
              GET >=>
                 choose [
                      route  "/"           >=> text "index" 
					  // ...
					  ]
				]) |> withConfig docsConfig

```

Since merge, __unit tests reveals a last bug.__
I need to investigate. I will try to fix it this week.


Regards,

**Comment by cotyar** [comment]
Many many thanks Flechner!!!

**Comment by dustinmoris** [comment]
Hi, I don't like letting people wait, but this is a bigger feature which needs more of my time to review so I wanted to give you a quick update on where I am:

I'm currently working on the 1.1.0 release which will have new router functionality and improved model binding to fix some of the issues which were sitting here for a while. I have tagged all issues with 1.1.0 and this release is scheduled to go out this week as soon as everything is done. Currently in the testing and documentation phase!

After that I've scheduled some improved auth features coming for 1.2.0. Mainly better CORS support in Giraffe (frequent request) and policy based auth to be on parity with MVC.

Once 1.2.0 is done I should have the time to get my head around this PR and the Swagger integration. Currently I am thinking that this should go into a separate repo: `Giraffe.Swagger`. Given where I am today it might take at least another couple weeks before I get to this point so if there is anything I can do today to help you please let me know.

**Comment by cotyar** [comment]
Guys, 
For the ones who don't want to wait until everything is done properly I just created a branch merging both PRs together (just as a temporal solution)
[https://github.com/cotyar/Giraffe/tree/SwaggerAndWebsockets](https://github.com/cotyar/Giraffe/tree/SwaggerAndWebsockets)
Cheers

**Comment by rflechner** [comment]
Hi,

I just pushed fixes.

You can now play with the sample :)

If you have time, could you try to find combination of route declaration causing a trouble of analysis ?

Regards,

**Comment by xperiandri** [comment]
So what is the progress with this PR?
Is it in process of review?

**Comment by cotyar** [comment]
@rflechner "I just pushed fixes." - merged your changes to the temporal [https://github.com/cotyar/Giraffe/tree/SwaggerAndWebsockets](https://github.com/cotyar/Giraffe/tree/SwaggerAndWebsockets)

**Comment by zaaack** [comment]
Would this support [TokenRouter](https://github.com/giraffe-fsharp/Giraffe/blob/master/DOCUMENTATION.md#tokenrouter)?

**Comment by rflechner** [comment]
I don't think but I could try to implement this in a next PR or commit.

**Comment by NatElkins** [comment]
So, I just tried to get this up and running on my machine (Ubuntu 18.04) and it seemed to work.  A few things though:

- The version of the dotnet SDK specified in the global.json is not available from the 18.04 repos.
image
Installing the `dotnet-sdk-2.1` package maps to a version number of `2.1.300`.  So I had to change that in the `global.json` to get this to build.

- The other thing is that it looks like the `dotnet-xunit` CLI tool needs to be installed at least once at some point before it can work.  I added a line like 
`<DotNetCliToolReference Include="dotnet-xunit" Version="2.3.1" />` (based on https://www.nuget.org/packages/dotnet-xunit/)  to `Giraffe.Tests.fsproj` and it started working.  Then when I switched back to master and tried to build again, it still worked.  So I think resolving the build issue is as simple as adding that line to the tests project.

Hope this helps.  Super excited to see this feature get in!!

File: tests/Giraffe.Tests/Giraffe.Tests.fsproj [hunk 1]
```diff
@@ -20,16 +18,14 @@
     <PackageReference Include="FsCheck" Version="3.0.0-*" />
     <PackageReference Include="FsCheck.Xunit" Version="3.0.0-*" />
   </ItemGroup>
-
   <ItemGroup>
-    <DotNetCliToolReference Include="dotnet-xunit" Version="2.3.*" />
```
Comment (NatElkins) [inline tests/Giraffe.Tests/Giraffe.Tests.fsproj]: I think this is why the tests are failing.

**Comment by ElijahReva** [comment]
@rflechner Could you give me access to this branch so I could fix it?

**Comment by ElijahReva** [comment]
@rflechner Thank you!

**Comment by rflechner** [comment]
@ElijahReva @NatElkins, is it right for you now ?

**Comment by NatElkins** [comment]
@rflechner Not able to check at the moment.  But while I've got you here, can I suggest making this feature an entirely separate repo/Nuget package?  That would allow it to have a separate release cadence from Giraffe and be more flexible about making updates to it.

When I last tested this feature with my project, an exception was thrown at startup (didn't have a chance to look too deeply into it).  I'll try to take another look and see what went wrong.

**Comment by NatElkins** [comment]
Also then @dustinmoris won't have to worry about it affecting the larger Giraffe code base.

**Comment by rflechner** [comment]
This PR is a technical proposition.
I think we should create another repository / nuget (in alpha) to be more agile.
Do you want to create a repo in your organization (https://github.com/giraffe-fsharp) ?

**Comment by cotyar** [comment]
Guys, @dustinmoris hasn't updated Giraffe since Feb.
Does anyone know if he's alright and if he still would like to continue supporting the project please?

**Comment by NatElkins** [comment]
@cotyar The develop branch was updated in May.

**Comment by czifro** [comment]
What's the best way to try this out while this PR is pending?

**Comment by rflechner** [comment]
I am closing this PR.
You can try the solution here: https://github.com/rflechner/SwaggerForFsharp

**Comment by dustinmoris** [comment]
https://github.com/giraffe-fsharp/Giraffe/issues/310

**Pull request closed without merging**